### PR TITLE
Kuznetsov Artyom lab 3 (option 1)

### PR DESCRIPTION
**Pull request by Kuznetsov-Artyom**
implemented a plugin that combines two instructions (`mul`/`add`) into one `muladd`

Example:
```cpp
// test.cpp
__m128d muladd_test2(__m128d a, __m128d b, __m128d c) {
  __m128d tmp = a * b;
  tmp = tmp + c;
  return tmp;
}
```
```ll
; test.ll
  define dso_local noundef <2 x double> @_Z12muladd_test2Dv2_dS_S_(<2 x double> noundef %a, <2 x double> noundef %b, <2 x double> noundef %c) local_unnamed_addr #2 {
  entry:
    %mul = fmul <2 x double> %a, %b
    %add = fadd <2 x double> %mul, %c
    ret <2 x double> %add
  }
  ```
```mir
# test_input.mir
  bb.0.entry:
    liveins: $xmm0, $xmm1, $xmm2
  
    renamable $xmm0 = nofpexcept MULPDrr killed renamable $xmm0, killed renamable $xmm1, implicit $mxcsr
    renamable $xmm0 = nofpexcept ADDPDrr killed renamable $xmm0, killed renamable $xmm2, implicit $mxcsr
    RET64 $xmm0
```
```mir
# test_outut.mir
iveins: $xmm0, $xmm1, $xmm2
  
    $xmm0 = VFMADD213PDZ128r $xmm0, $xmm1, $xmm2, implicit $mxcsr
    RET64 $xmm0
```

### CODE REVIEW

File: llvm/test/lab3/kuznetsov_artyom/testing.mir [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+# __m128d muladd_test2(__m128d a, __m128d b, __m128d c) {
+#   __m128d tmp = a * b;
+#   tmp = tmp + c;
+#   return tmp;
+# }
+
+# __m128d muladd_test3(__m128d a, __m128d b, __m128d c) {
+#   __m128d tmp = a * c + b;
+#   return tmp * c + b;
+# }
```
Comment (m-ly4) [inline llvm/test/lab3/kuznetsov_artyom/testing.mir]:
```suggestion
# __m128d muladd_test3(__m128d a, __m128d b, __m128d c) {
#   __m128d tmp = a * c + b;
#   return tmp * c + b;
# }

# __m128d muladd_test4(__m128d a, __m128d b, __m128d c) {
#   __m128d tmp = a * c;
#   tmp = tmp / b;
#   return tmp + b;
# }

# __m128d muladd_test5(__m128d a, __m128d b, __m128d c) {
#   __m128d tmp = a * b;
#   __m128d tmp2 = b + a -c;
#   return tmp + c + tmp2;
# }

# __m128d muladd_test6(__m128d a, __m128d b, __m128d c) {
#   __m128d tmp = a * b;
#   __m128d tmp2 = tmp + c;
#   return tmp * c;
# }
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kuznetsov_artyom/testing.mir (hunk 1 above)
Comment (Kuznetsov-Artyom) [reply llvm/test/lab3/kuznetsov_artyom/testing.mir]: @m-ly4 I have finalized my pass. Thanks for the provided cases! 🔥

File: llvm/lib/Target/X86/X86.h [hunk 2]
```diff
@@ -201,6 +202,7 @@ void initializeX86ReturnThunksPass(PassRegistry &);
 void initializeX86SpeculativeExecutionSideEffectSuppressionPass(PassRegistry &);
 void initializeX86SpeculativeLoadHardeningPassPass(PassRegistry &);
 void initializeX86TileConfigPass(PassRegistry &);
+void initializeX86MulAddPassPass(PassRegistry &);
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86.h]: Please rewrite plugin as dyn lib. You may look into #99 as example.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/X86.h (hunk 2 above)
Comment (Kuznetsov-Artyom) [reply llvm/lib/Target/X86/X86.h]: > Please rewrite plugin as dyn lib. You may look into #99 as example.

done ✅

File: llvm/lib/Target/X86/lab3/kuznetsov_artyom/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+set(PluginName X86MulAddPass)
+file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h)
+add_llvm_pass_plugin(${PluginName}
+        ${ALL_SOURCE_FILES}
+        DEPENDS
+        intrinsics_gen
+        X86
+        BUILDTREE_ONLY)
+
+target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kuznetsov_artyom/CMakeLists.txt:10]:
```suggestion
target_include_directories(${PluginName} PUBLIC ${PATH_TO_X86})

set(LLVM_TEST_DEPENDS ${PluginName} ${LLVM_TEST_DEPENDS} PARENT_SCOPE)

```

**Review by m-ly4 — APPROVED** [review]

File: llvm/lib/Target/X86/lab3/kuznetsov_artyom/CMakeLists.txt (hunk 3 above)
Comment (Kuznetsov-Artyom) [reply llvm/lib/Target/X86/lab3/kuznetsov_artyom/CMakeLists.txt:10]: P.S : There is no need to add this dependency, when building your plugin, it is enough to prescribe `cmake --build . --target NamePass`, where we specify `NamePass` in `CMakeLists.txt` . And in the pool revquest, the project builds completely, so our targets are automatically picked up

Comment (Kuznetsov-Artyom) [reply llvm/lib/Target/X86/lab3/kuznetsov_artyom/CMakeLists.txt:10]: I've double-checked it now, even if I write `cmake --build . --target check-llvm-lab3` on my system, then the changes made to the file `X86MulAddPass.cpp` they don't catch on. The same will happen with `cmake --build . --target check-llvm`

UPD: I fixed one moment, everything worked 👍

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**